File: python/counter_race_condition.py
from threading import Thread

counter = 0

def increment_counter_by(n):
    global counter
    for i in range(n):
        counter += 1


def main():
    t1 = Thread(target=increment_counter_by, args=(1000000, ))
    t2 = Thread(target=increment_counter_by, args=(1000000, ))

    t1.start()
    t2.start()

    t1.join()
    t2.join()

    print(f"Counter: {counter}")


if __name__ == '__main__':
    main()

File: cpp/counter_race_condition.cpp
#include <thread>
#include <iostream>

// en que segmento de memoria está esta variable?
// los threads posix comparten este segmento de memoria?
int counter = 0;


void increment_counter_by(int n) {
    for (int i = 0; i < n; ++i) {
        counter++;
    }
}

int main() {
    // spawn threads
    std::thread t1(increment_counter_by, 1000000);
    std::thread t2(increment_counter_by, 1000000);
    // join threads
    t1.join();
    t2.join();

    std::cout << "Counter: " << counter << std::endl;
    return 0;
}